### PR TITLE
Add support for delegate GDV and method-based vtable GDV

### PR DESCRIPTION
**Pull request by jakobbotsch**
Add support for instrumenting delegate calls and vtable calls into method handle histograms. Use these histograms to do GDV for delegate calls and also support method-based GDV for vtable calls.

For instrumentation we now support class probes at interface call sites, method probes at delegate call sites and both class probes and method probes at vtable call sites. For vtable calls, when turned on, instrumentation produces both histograms as PGO data so that the JIT can later make the choice about what is the best form of guard to use at that site.

For guarding, there are some things to take into account. Delegate calls currently (practically) always point to precode, so this PR is just guarding on `getFunctionFixedEntryPoint` which returns the precode address, and this is generally quite cheap (same cost as class-based GDV). That's the case for delegates pointing to instance methods anyway, this PR does not support static methods yet -- those will be more expensive.

For vtable calls the runtime will backpatch the slots when tiering, so the JIT guards the address retrieved from the vtable against an indirection of the slot, which is slightly more expensive than a class-based guard.

Currently the instrumentation is enabled conditionally with `COMPlus_JitDelegateProfiling=1` (for delegates) and `COMPlus_JitVTableProfiling=1` (for vtable calls). Currently delegate profiling is turned on by default while vtable profiling is off by default.
Simple microbenchmark:
```csharp
public class Benchmark
{
    private readonly long[] _nums;
    public Benchmark()
    {
        _nums = Enumerable.Range(0, 100000).Select(i => (long)i).ToArray();
    }

    [Benchmark]
    public long Sum() => _nums.Sum(l => l * l);
}
```

| Method |        Job |                              Toolchain |      Mean |    Error |   StdDev | Ratio |
|------- |----------- |--------------------------------------- |----------:|---------:|---------:|------:|
|    Sum | Job-QWNDLL |                     \nopgo\corerun.exe | 406.65 us | 0.718 us | 0.560 us |  1.00 |
|    Sum | Job-PNPEDU |                 \tieredpgo_no_delegate_gdv\corerun.exe | 172.77 us | 0.819 us | 0.766 us |  0.42 |
|    Sum | Job-KFFWQK | \tieredpgo_delegate_gdv\corerun.exe |  91.38 us | 0.263 us | 0.219 us |  0.22 |

TODO:
* [X] Work out why chained GDV regresses block layout sometimes (resolved with https://github.com/dotnet/runtime/pull/69878)
* [x] Add to PGO CI pipeline
* [x] Support randomized GDVs for method handle histograms
* [x] Investigate some OSR interactions
* [x] Evaluate proper likelihood heuristics for delegates
* [x] Optimize cold path for delegate GDV -- we should reuse the target
* [x] Devirtualized delegate calls should not null check
* [x] Optimize profiling helpers
* [x] Add signature compatibility check
* [x] Add method address handle comment in disasm (https://github.com/dotnet/runtime/pull/70437)
* [x] Separate ability to do method profiling for delegates and vtable calls
* [ ] ~~Support delegates pointing to value type instance methods~~ (Follow-up)
* [ ] ~~Support delegates pointing to static functions~~ (Follow-up)
* [ ] ~~R2R support~~ (Follow-up)

Contributes to #44610

cc @dotnet/jit-contrib

### CODE REVIEW

**Comment by bencyoung-Fignum** [comment]
This is amazing. In theory could it hoist the check, clone the loop and then have a loop with a simple add, and a loop with the fallback?

**Comment by AndyAyersMS** [comment]
> This is amazing. In theory could it hoist the check, clone the loop and then have a loop with a simple add, and a loop with the fallback?

Yes, we're working on that too. See https://github.com/dotnet/runtime/issues/65206 which includes a link to a prototype that does this (at least for type tests -- should not be hard to make it work for delegate tests as well).

**Comment by AndyAyersMS** [comment]
Very cool to see this!  I just skimmed the changes; will take a closer look -- may not get to it until next week.

We should push to get the prereq https://github.com/dotnet/runtime/pull/67919 merged.

> There are some unexpected layout issues due to chained GDV, for example in simple examples `Enumerable.Sum` regresses when method profiling is on because the JIT sees an extra chained GDV opportunity and ends up making a mess of the flow graph.

Not sure what kind of messes you're seeing, but GDV even without chaining can cause problems, eg https://github.com/dotnet/runtime/issues/67318.

> so that the JIT can later make the choice about what is the best form of guard to use at that site.

Would be interesting to see how many sites are type-polymorphic but method-monomorphic.

> both class probes and method probes at vtable call sites.

We should probably work on streamlining the cost of the class/method probes. Simple thing would be to hoist up the check where we decide to update the table; if we're not going to do a table update we don't need to do all the work to figure out what values we'd record. That should help somewhat, especially for the class/method version where those computations are more complex.

**Comment by jakobbotsch** [comment]
> We should probably work on streamlining the cost of the class/method probes. Simple thing would be to hoist up the check where we decide to update the table; if we're not going to do a table update we don't need to do all the work to figure out what values we'd record. That should help somewhat, especially for the class/method version where those computations are more complex.

That optimization makes sense. On another note I'm not sure if the combined profile helper is worth much, since from the call site and class we should be able to get the resolved method anyway. So it might just go away in the final version.

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -9536,13 +9536,20 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
             //
             methHnd = callInfo->hMethod;
         }
+        else if (call->AsCall()->IsDelegateInvoke() && opts.OptimizationEnabled() && !opts.IsOSR())
+        {
+            considerGuardedDevirtualization(call->AsCall(), rawILOffset, false, NO_METHOD_HANDLE, NO_CLASS_HANDLE,
+                                            nullptr);
+        }
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:9543]: @AndyAyersMS I was seeing issues where we try to both do GDV and add probes in OSR variants. There are comments that say that we currently do not devirtualize in OSR variants here:
https://github.com/dotnet/runtime/blob/66bd8b193bdaa96e26308e6c056f0b27434428cd/src/coreclr/jit/importer.cpp#L20909-L20912

Do you know off the top of your head where we ensure that no GDV should happen for OSR variants?

Comment (jakobbotsch) [reply src/coreclr/jit/importer.cpp:9543]: Ah, I think it is the early out of `impDevirtualizeCall` that I have mistakenly removed while extracting `impConsiderCallProbe`.

Comment (AndyAyersMS) [reply src/coreclr/jit/importer.cpp]: Yeah, looks like you found it.

**Comment by jakobbotsch** [comment]
> Not sure what kind of messes you're seeing, but GDV even without chaining can cause problems, eg #67318.

Indeed, the case I was seeing look fixed now after #69878. Simple `Enumerable.Sum` benchmark is now significantly faster:
```csharp
public class Benchmark
{
    private readonly long[] _nums;
    public Benchmark()
    {
        _nums = Enumerable.Range(0, 100000).Select(i => (long)i).ToArray();
    }

    [Benchmark]
    public long Sum() => _nums.Sum(l => l * l);
}
```
Results:

| Method |        Job |                              Toolchain |      Mean |    Error |   StdDev | Ratio |
|------- |----------- |--------------------------------------- |----------:|---------:|---------:|------:|
|    Sum | Job-QWNDLL |                     \nopgo\corerun.exe | 406.65 us | 0.718 us | 0.560 us |  1.00 |
|    Sum | Job-PNPEDU |                 \tieredpgo\corerun.exe | 172.77 us | 0.819 us | 0.766 us |  0.42 |
|    Sum | Job-KFFWQK | \tieredpgo_methodprofiling\corerun.exe |  91.38 us | 0.263 us | 0.219 us |  0.22 |

**Comment by jakobbotsch** [comment]
I have two test runs where I raised the timeout significantly so that they completed. They are:
runtime-coreclr libraries-pgo: https://dev.azure.com/dnceng/public/_build/results?buildId=1802466&view=results
runtime-coreclr pgo: https://dev.azure.com/dnceng/public/_build/results?buildId=1802465&view=results

There are a few failures but the only failure in the methodprofiling configuration looks to be #70231.

**Comment by jakobbotsch** [comment]
I do not see much impact in either direction on TechEmpower:
image

That's not that unexpected, we do not use delegates very much in those benchmarks. Here's the data in terms of how many times a probe at a specific IL offset was hit:
https://gist.github.com/jakobbotsch/1ef169922db9be5b3740e05a5ea9544f
For example, the data shows that the function in database-fortunes that has the most collected delegate GDV data was `Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.IOQueue.System.Threading.IThreadPoolWorkItem.Execute`, which had a delegate call site at IL offset 27 that was probed 150243 times (I increased the call counting threshold when gathering the data).

The same data for type GDV is https://gist.github.com/jakobbotsch/0903b426eddb69fd830662b228d5ca32
which has way more data (~12x more lines in that text output) and also the call sites are much hotter.

**Comment by jakobbotsch** [comment]
This should be ready. I am enabling delegate GDV by default in this PR, while keeping the vtable profiling disabled. The impact on size of PGO data and SPC.dll with delegate GDV is:

| **File**                      | **Before** | **After** | **PDIFF** |
|-------------------------------|------------|-----------|-----------|
| DotNet_FirstTimeXP.mibc       | 2538 KB    | 2570 KB   | +1.26%    |
| DotNet_HelloWorld.mibc        | 5239 KB    | 5263 KB   | +0.46%    |
| DotNet_OrchardCore.mibc       |  2339 KB   | 2383 KB   | +1.88%    |
| DotNet_TechEmpower.mibc       | 976 KB     | 992 KB    | +1.64%    |
| StandardOptimizationData.mibc | 7377 KB    | 7494 KB   | +1.59%    |
| System.Private.CoreLib.dll        | 11168 KB   | 11168 KB  | +0.00%    |

For TechEmpower impact on PGO, see the comment above.

cc @dotnet/jit-contrib PTAL @AndyAyersMS @EgorBo

File: eng/pipelines/common/templates/runtimes/run-test-job.yml [hunk 2]
```diff
@@ -227,13 +227,13 @@ jobs:
...
-    ${{ if in(parameters.testGroup, 'outerloop', 'jit-experimental', 'pgo', 'jit-cfg') }}:
+    ${{ if in(parameters.testGroup, 'outerloop', 'jit-experimental', 'jit-cfg') }}:
       timeoutInMinutes: 270
     ${{ if in(parameters.testGroup, 'gc-longrunning', 'gc-simulator') }}:
       timeoutInMinutes: 480
     ${{ if in(parameters.testGroup, 'jitstress', 'jitstress-isas-arm', 'jitstressregs-x86', 'jitstressregs', 'jitstress2-jitstressregs', 'gcstress0x3-gcstress0xc', 'ilasm') }}:
       timeoutInMinutes: 390
-    ${{ if in(parameters.testGroup, 'gcstress-extra', 'r2r-extra', 'clrinterpreter') }}:
+    ${{ if in(parameters.testGroup, 'gcstress-extra', 'r2r-extra', 'clrinterpreter', 'pgo') }}:
       timeoutInMinutes: 510
```
Comment (jakobbotsch) [inline eng/pipelines/common/templates/runtimes/run-test-job.yml:237]: I am thinking to keep these timeout changes until we get around to splitting up the jobs, it fixes the problem we are seeing with timeouts and I have not seen any issues with making the timeout this high.

File: src/coreclr/jit/importer.cpp [hunk 3]
```diff
@@ -21389,51 +21405,7 @@ void Compiler::impDevirtualizeCall(GenTreeCall*            call,
...
-            compCurBB->bbFlags |= BBF_HAS_CLASS_PROFILE;
-        }
-        return;
-    }
-
-    // Bail if optimizations are disabled.
-    if (opts.OptimizationDisabled())
-    {
-        return;
-    }
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:21436]: This code was extracted into impConsiderCallProbe which is called in impImportCall instead. Unlike impDevirtualizeCall, impConsiderCallProbe is called for delegates too.

File: src/coreclr/jit/importer.cpp [hunk 4]
```diff
@@ -22225,153 +22305,378 @@ void Compiler::addFatPointerCandidate(GenTreeCall* call)
...
-    CORINFO_DEVIRTUALIZATION_INFO dvInfo;
-    dvInfo.virtualMethod               = baseMethod;
-    dvInfo.objClass                    = likelyClass;
-    dvInfo.context                     = *pContextHandle;
-    dvInfo.exactContext                = *pContextHandle;
-    dvInfo.pResolvedTokenVirtualMethod = nullptr;
+    uint32_t likelyClassAttribs = 0;
+    if (likelyClass != NO_CLASS_HANDLE)
+    {
+        likelyClassAttribs = info.compCompHnd->getClassAttribs(likelyClass);
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:22607]: This stuff is easier to review with whitespace ignored.

**Comment by jakobbotsch** [comment]
Test failure is #68376. superpmi failures are expected due to JIT-EE GUID change.

**Review by AndyAyersMS — COMMENTED** [review]
Overall this looks great.

File: src/coreclr/jit/compiler.h [hunk 5]
```diff
@@ -6795,13 +6811,26 @@ class Compiler
         optMethodFlags |= OMF_HAS_GUARDEDDEVIRT;
     }
 
+    void clearMethodHasGuardedDevirtualization()
+    {
+        optMethodFlags &= ~OMF_HAS_GUARDEDDEVIRT;
+    }
+
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/compiler.h:6818]: Ah, I was wondering why I added this. Looks like a bad merge. Will fix.

File: src/coreclr/jit/fgprofile.cpp [hunk 6]
```diff
@@ -1440,89 +1440,101 @@ class ClassProbeVisitor final : public GenTreeVisitor<ClassProbeVisitor<TFunctor
...
+        }
+
+        if ((probeType == Compiler::GDVProbeType::MethodProfile) ||
+            (probeType == Compiler::GDVProbeType::MethodAndClassProfile))
+        {
+            CreateHistogramEntriesIfNecessary(compiler, call, false /* isTypeHistogram */);
+        }
+    }
+
+    void CreateHistogramEntriesIfNecessary(Compiler* compiler, GenTreeCall* call, bool isTypeHistogram)
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/fgprofile.cpp:1492]: Hmm, I think at one point I was checking the probe type in that function but decided to do the current thing. Will rename it.

**Comment by jakobbotsch** [comment]
Also PTAL @davidwrighton for the non-JIT/SPMI changes. I have bumped the R2R minor version again since we now may see `GetLikelyMethod` entries in the embedded PGO data.

One thing I would like to point attention to is that old versions of dotnet-pgo (from before #67919) fail with an "Unknown PGO type" error if they see this new data. So one needs to use a version of dotnet-pgo that matches the version of the runtime used to produce the data.

**Review by davidwrighton — APPROVED** [review]
:shipit:

**Review by AndyAyersMS — APPROVED** [review]

**Comment by davidfowl** [comment]
I missed this change. This is a big deal 😀 !

**Pull request merged**